File: sandbox-runner/patterns/mapreduce-tree/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { RECORDS, TASK_QUEUE, WORKFLOW_ID_PREFIX } from "./shared";
import { nodeWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const workflowId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;
    const handle = await client.workflow.start(nodeWorkflow, {
      args: [RECORDS, 0, ""],
      taskQueue: TASK_QUEUE,
      workflowId,
    });
    console.log(`Started workflow: ${workflowId}`);
    console.log(`Processing ${RECORDS.length} records via MapReduce Tree…`);

    const results = await handle.result();
    console.log(`MapReduce Tree complete: ${results.length} results`);
    console.log(`Results: ${results.join(", ")}`);
    console.log(
      `Open the Temporal UI and search for '${workflowId}' to see the full workflow tree.`
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sandbox-runner/patterns/mapreduce-tree/python/starter.py
import asyncio
import time

from temporalio.client import Client

from shared import RECORDS, TASK_QUEUE, WORKFLOW_ID_PREFIX
from workflows import NodeWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")
    workflow_id = f"{WORKFLOW_ID_PREFIX}-{int(time.time() * 1000)}"
    handle = await client.start_workflow(
        NodeWorkflow.run,
        args=[RECORDS, 0, ""],
        id=workflow_id,
        task_queue=TASK_QUEUE,
    )
    print(f"Started workflow: {workflow_id}")
    print(f"Processing {len(RECORDS)} records via MapReduce Tree…")

    results = await handle.result()
    print(f"MapReduce Tree complete: {len(results)} results")
    print(f"Results: {', '.join(results)}")
    print(
        f"Open the Temporal UI and search for '{workflow_id}' to see the full workflow tree."
    )


if __name__ == "__main__":
    asyncio.run(main())
